python_modules/dagster/dagster/_cli/definitions.py

```diff
@@ -0,0 +1,68 @@
+import sys
+
+import click
+
+from dagster._core.origin import DEFAULT_DAGSTER_ENTRY_POINT
+from dagster._core.remote_representation import ManagedGrpcPythonEnvCodeLocationOrigin
+from dagster._grpc.server import LoadedRepositories
+
+from .job import apply_click_params
+from .workspace.cli_target import (
+    ClickArgValue,
+    get_workspace_load_target,
+    python_file_option,
+    python_module_option,
+    workspace_option,
+)
+
+
+@click.group(name="definitions")
+def definitions_cli():
+    """Commands for working with Dagster definitions."""
+
+
+def validate_command_options(f):
+    return apply_click_params(
+        f,
+        workspace_option(),
+        python_file_option(allow_multiple=True),
+        python_module_option(allow_multiple=True),
+    )
+
+
+@validate_command_options
+@definitions_cli.command(
+    name="validate",
+    help="Validate if Dagster definitions are loadable.",
+)
+def definitions_validate_command(**kwargs: ClickArgValue):
+    workspace_load_target = get_workspace_load_target(kwargs)
+    code_locations_origins = workspace_load_target.create_origins()
+
+    click.echo("Starting validation...")
+    for code_location_origin in code_locations_origins:
+        # CodeLocationOrigin objects are of type GrpcServerCodeLocationOrigin
+        # when loading from a grpc server in workspace.yaml.
+        if not isinstance(code_location_origin, ManagedGrpcPythonEnvCodeLocationOrigin):
+            continue
+
+        curr_target_origin = (
+            code_location_origin.loadable_target_origin.python_file
+            or code_location_origin.loadable_target_origin.module_name
+            or code_location_origin.loadable_target_origin.package_name
+        )
+        click.echo(f"Validating definitions in {curr_target_origin}.")
+        try:
+            LoadedRepositories(
+                loadable_target_origin=code_location_origin.loadable_target_origin,
+                entry_point=DEFAULT_DAGSTER_ENTRY_POINT,
+            )
+        except Exception as e:
+            click.echo(
+                click.style("Validation failed with exception: ", fg="red") + f"{e}.", err=True
+            )
+            sys.exit(1)
+        else:
+            click.echo("Validation successful!")
```

Comment:
One thing that I don't like about this is that it loads all the different code locations in the same process. It's not a given that if you have two different code locations that they will play nicely together - they might have inconsistent packages that don't work together for example.

Would it be possible to load each of these up in a subprocess like the webserver does on startup? And pass along the load error over gRPC for any of them if there are any?

Reply:
That makes perfect sense - done in [655b3a6](https://github.com/dagster-io/dagster/pull/23692/commits/655b3a651eaa62f36d2784cbdb21a3fb06d0f748)